File: cpp-answers/257.二叉树的所有路径.cpp
/*
 * @Author: your name
 * @Date: 2020-06-13 19:11:25
 * @LastEditTime: 2020-06-13 19:18:47
 * @LastEditors: Please set LastEditors
 * @Description: In User Settings Edit
 * @FilePath: \undefinede:\leetcode刷题\257.二叉树的所有路径.cpp
 */
/*
 * @lc app=leetcode.cn id=257 lang=cpp
 *
 * [257] 二叉树的所有路径
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution
{
public:
    vector<string> binaryTreePaths(TreeNode *root)
    {
        vector<string> v;
        if (root)
            preorder(v, root, "");
        return v;
    }
    void preorder(vector<string> &v, TreeNode *r, string t)
    {
        if (!r)
            return;
        if (!t.empty())
            t += ("->" + to_string(r->val));
        else
            t += to_string(r->val);
        if (r->left || r->right)
        {
            preorder(v, r->left, t);
            preorder(v, r->right, t);
        }
        else
            v.push_back(t);
    }
};
// @lc code=end

File: rust-answers/257.二叉树的所有路径.rs
/*
 * @lc app=leetcode.cn id=257 lang=rust
 *
 * [257] 二叉树的所有路径
 */

// @lc code=start
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
type OptNode = Option<Rc<RefCell<TreeNode>>>;

impl Solution {
    pub fn binary_tree_paths(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<String> {
        let mut ans = Vec::new();
        Self::dfs(&root, &mut Vec::new(), &mut ans);
        ans
    }

    fn dfs(node: &OptNode, curr: &mut Vec<String>, ans: &mut Vec<String>) {
        if let Some(n) = node.as_ref() {
            let b = n.borrow();
            curr.push(b.val.to_string());
            if b.left.is_none() && b.right.is_none() {
                ans.push(curr.join("->"));
            } else {
                Self::dfs(&b.left, curr, ans);
                Self::dfs(&b.right, curr, ans);
            }
            curr.pop();
        }
    }
}
// @lc code=end

